### PR TITLE
Fix enable and disable scheduler order on upgrade

### PR DESCRIPTION
**Pull request by gabrielcocenza**
- disable the nova-compute scheduler should happen before upgrading any colocated application. E.g: It's not good practice to start upgrading cinder, colocated with nova-compute,  without disabling the scheduler before
- enable the nova-compute scheduler should happen after all apps of the hypervisor are done. E.g: It's not good practice enable the nova-compute scheduler if cinder, colocated with nova-compute, didn't upgrade yet.

Fix: #299

### CODE REVIEW

File: cou/apps/core.py
Comment (agileshaw) [inline cou/apps/core.py:120]: non-blocking: I think it would be better to add a new class `UnitUpgradePlan` to be used here.

File: cou/apps/core.py
Comment (agileshaw) [inline cou/apps/core.py:168]:
```suggestion
        """Get the step to disable the scheduler, so the unit cannot create new VMs.
```

**Review by agileshaw — COMMENTED** [review]
The changes LGTM. But I think we should wait until #339 is merged to `main` (or rebase on it) so that we can update the documentation also in this PR. Otherwise it's easy to forget.

File: cou/apps/core.py
Comment (rgildein) [reply cou/apps/core.py:120]: I agree that that would be nice, it's not related with this PR, so I will suggest to no to it here. We can report an issue about this.

not-blocking: Why you are adding `description=` back? I do not see any benefit of it and in a lot of places I tried to remove it, so make lines shorter.

File: cou/apps/core.py
Comment (rgildein) [inline cou/apps/core.py:156]: not-blocking: 
```suggestion
            units = self.units.values()
```

File: cou/apps/core.py
Comment (rgildein) [inline cou/apps/core.py:155]: We are using same if condition in multiple places, but now I realized that it maybe an issue if I want to generate upgrade plan for app but not touching any units, so I passed `units == []`. Should we rather use `if units is None`? What do you think?

File: cou/steps/hypervisor.py
Comment (rgildein) [inline cou/steps/hypervisor.py:240]: I still think that running in reverse order make more sense, but if you think that it can be harmful, I'm ok to keep this change.

File: tests/unit/apps/test_core.py
Comment (rgildein) [inline tests/unit/apps/test_core.py:800]:
```suggestion
    units_selected = [app.units[unit] for unit in units]  # units instead of units_slice, where units is list of names
```

File: cou/apps/core.py
Comment (rgildein) [inline cou/apps/core.py:181]: not-blocking: What if we use one functions with `enable=True` for enabling (run enable action) and `enable=False` for disabling (run disable action).

**Review by rgildein — CHANGES_REQUESTED** [review]

File: cou/apps/core.py
Comment (rgildein) [reply cou/apps/core.py:181]: Good point, resolving. :+1:

File: cou/apps/core.py
Comment (gabrielcocenza) [reply cou/apps/core.py:155]: Interesting...
I think passing `units = []` will only happen on all units of the hypervisor is upgraded and then `_check_application_target` will halt the upgrade plan generation. Howevwer, we can have a single place to check if units is None and not checking every time.

I will add this is this patch. It's a little bit out of the scope of this PR, but I don't think it's a big deal

Comment (rgildein) [reply cou/apps/core.py:155]: What do you mean by single place to check it?

Comment (gabrielcocenza) [reply cou/apps/core.py:155]: Digging a little bit it looks like that the changes required are more complex than I expected so I'll make it simple in this PR as you suggested. I've opened a issue #344 so we can better think about this

File: cou/steps/hypervisor.py
Comment (gabrielcocenza) [reply cou/steps/hypervisor.py:240]: To say the truth I don't understand why we would change the order.
1. We are not using the hypervisor to upgrade subordinates, so the description is outdated
2. Because of the upgrade order, control-plane colocated apps starts to upgrade first and then nova-compute

I think we should keep the way it is and I don't see harm

File: cou/apps/core.py
Comment (agileshaw) [reply cou/apps/core.py:120]: Issue for adding `UnitUpgradePlan` created: https://github.com/canonical/charmed-openstack-upgrader/issues/346

File: cou/apps/core.py
Comment (gabrielcocenza) [reply cou/apps/core.py]: > What do you mean by single place to check it?

As you said: `We are using same if condition in multiple places` That is what I meant by single place to check that condition.
This condition is happening because we need to check in multiple functions which upgrade strategy is used. We should think a better and more elegant solution for this later.

File: tests/unit/apps/test_core.py
Comment (rgildein) [inline tests/unit/apps/test_core.py:842]: not-blocking: I think it would be ok to define without range, but definitely no blocker.

**Review by rgildein — APPROVED** [review]

**Review by rgildein — APPROVED** [review]
lgtm

**Review by agileshaw — APPROVED** [review]

**Pull request merged**
